File: jbake-core/src/main/java/org/jbake/app/configuration/JBakeConfigurationFactory.java
package org.jbake.app.configuration;


import org.apache.commons.configuration2.CompositeConfiguration;
import org.apache.commons.configuration2.ex.ConfigurationException;

import java.io.File;

/**
 * A {@link JBakeConfiguration} factory
 */
public class JBakeConfigurationFactory {

    private ConfigUtil configUtil;

    public JBakeConfigurationFactory() {
        this.configUtil = new ConfigUtil();
    }

    /**
     * Creates a {@link DefaultJBakeConfiguration}
     * @param sourceFolder The source folder of the project
     * @param destination  The destination folder to render and copy files to
     * @param isClearCache Whether to clear database cache or not
     * @return A configuration by given parameters
     * @throws ConfigurationException if loading the configuration fails
     * @deprecated use {@link #createDefaultJbakeConfiguration(File, File, File, boolean)} instead
     */
    @Deprecated
    public DefaultJBakeConfiguration createDefaultJbakeConfiguration(File sourceFolder, File destination, boolean isClearCache) throws ConfigurationException {
        return createDefaultJbakeConfiguration(sourceFolder, destination, (File) null, isClearCache);
    }

    /**
     * Creates a {@link DefaultJBakeConfiguration}
     * @param sourceFolder   The source folder of the project
     * @param destination    The destination folder to render and copy files to
     * @param propertiesFile The properties file for the project
     * @param isClearCache   Whether to clear database cache or not
     * @return A configuration by given parameters
     * @throws ConfigurationException if loading the configuration fails
     */
    public DefaultJBakeConfiguration createDefaultJbakeConfiguration(File sourceFolder, File destination, File propertiesFile, boolean isClearCache) throws ConfigurationException {
        DefaultJBakeConfiguration configuration = (DefaultJBakeConfiguration) getConfigUtil().loadConfig(sourceFolder, propertiesFile);
        configuration.setDestinationFolder(destination);
        configuration.setClearCache(isClearCache);
        return configuration;
    }

    /**
     * Creates a {@link DefaultJBakeConfiguration}
     * This is a compatibility factory method
     *
     * @param sourceFolder           The source folder of the project
     * @param destination            The destination folder to render and copy files to
     * @param compositeConfiguration A given {@link CompositeConfiguration}
     * @param isClearCache           Whether to clear database cache or not
     * @return A configuration by given parameters
     */
    public DefaultJBakeConfiguration createDefaultJbakeConfiguration(File sourceFolder, File destination, CompositeConfiguration compositeConfiguration, boolean isClearCache) {
        DefaultJBakeConfiguration configuration = new DefaultJBakeConfiguration(sourceFolder, compositeConfiguration);
        configuration.setDestinationFolder(destination);
        configuration.setClearCache(isClearCache);
        return configuration;
    }

    /**
     * Creates a {@link DefaultJBakeConfiguration}
     * This is a compatibility factory method
     *
     * @param sourceFolder           The source folder of the project
     * @param destination            The destination folder to render and copy files to
     * @param compositeConfiguration A given {@link CompositeConfiguration}
     * @return A configuration by given parameters
     */
    public DefaultJBakeConfiguration createDefaultJbakeConfiguration(File sourceFolder, File destination, CompositeConfiguration compositeConfiguration) {
        DefaultJBakeConfiguration configuration = new DefaultJBakeConfiguration(sourceFolder, compositeConfiguration);
        configuration.setDestinationFolder(destination);
        return configuration;
    }

    /**
     * Creates a {@link DefaultJBakeConfiguration}
     *
     * @param sourceFolder The source folder of the project
     * @param config       A {@link CompositeConfiguration}
     * @return A configuration by given parameters
     */
    public DefaultJBakeConfiguration createDefaultJbakeConfiguration(File sourceFolder, CompositeConfiguration config) {
        return new DefaultJBakeConfiguration(sourceFolder, config);
    }

    /**
     * Creates a {@link DefaultJBakeConfiguration} with value site.host replaced
     * by http://localhost:[server.port].
     * The server.port is read from the project properties file.
     *
     * @param sourceFolder      The source folder of the project
     * @param destinationFolder The destination folder to render and copy files to
     * @param isClearCache      Whether to clear database cache or not
     * @return A configuration by given parameters
     * @throws ConfigurationException if loading the configuration fails
     * @deprecated use {@link #createJettyJbakeConfiguration(File, File, File, boolean)} instead
     */
    @Deprecated
    public DefaultJBakeConfiguration createJettyJbakeConfiguration(File sourceFolder, File destinationFolder, boolean isClearCache) throws ConfigurationException {
        return createJettyJbakeConfiguration(sourceFolder, destinationFolder, (File) null, isClearCache);
    }

    /**
     * Creates a {@link DefaultJBakeConfiguration} with value site.host replaced
     * by http://localhost:[server.port].
     * The server.port is read from the project properties file.
     *
     * @param sourceFolder      The source folder of the project
     * @param destinationFolder The destination folder to render and copy files to
     * @param propertiesFile    The properties file for the project
     * @param isClearCache      Whether to clear database cache or not
     * @return A configuration by given parameters
     * @throws ConfigurationException if loading the configuration fails
     */
    public DefaultJBakeConfiguration createJettyJbakeConfiguration(File sourceFolder, File destinationFolder, File propertiesFile, boolean isClearCache) throws ConfigurationException {
        DefaultJBakeConfiguration configuration = (DefaultJBakeConfiguration) getConfigUtil().loadConfig(sourceFolder, propertiesFile);
        configuration.setDestinationFolder(destinationFolder);
        configuration.setClearCache(isClearCache);
        configuration.setSiteHost("http://" + configuration.getServerHostname() + ":" + configuration.getServerPort() + configuration.getServerContextPath());
        return configuration;
    }

    public ConfigUtil getConfigUtil() {
        return configUtil;
    }

    public void setConfigUtil(ConfigUtil configUtil) {
        this.configUtil = configUtil;
    }

    public JBakeConfigurationFactory setEncoding(String charset) {
        this.configUtil.setEncoding(charset);
        return this;
    }
}

File: jbake-core/src/main/java/org/jbake/app/configuration/JBakeConfigurationFactory.kt
package org.jbake.app.configuration

import org.jbake.app.JBakeExitException
import java.io.File


/**
 * A [JBakeConfiguration] factory.
 */
class JBakeConfigurationFactory {

    @JvmField var configUtil: ConfigUtil = ConfigUtil()

    /**
     * Creates a [DefaultJBakeConfiguration] using default.properties and jbake.properties
     *
     * @param sourceDir The source directory of the project
     * @param destination The destination directory to render and copy files to
     * @param isClearCache Whether to clear database cache or not
     * @throws JBakeExitException if loading the configuration fails
     */
    @Throws(JBakeExitException::class)
    fun createDefaultJbakeConfiguration(sourceDir: File, destination: File, isClearCache: Boolean): DefaultJBakeConfiguration
        = createDefaultJbakeConfiguration(sourceDir, destination, null as File?, isClearCache)

    /**
     * Creates a [DefaultJBakeConfiguration]
     * @param sourceDir The source directory of the project
     * @param destination The destination directory to render and copy files to
     * @param propertiesFile The properties file for the project
     * @param isClearCache Whether to clear database cache or not
     * @throws JBakeExitException if loading the configuration fails
     */
    @Throws(JBakeExitException::class)
    fun createDefaultJbakeConfiguration(
        sourceDir: File,
        destination: File,
        propertiesFile: File?,
        isClearCache: Boolean,
    ): DefaultJBakeConfiguration {
        val configuration = this.configUtil.loadConfig(sourceDir, propertiesFile) as DefaultJBakeConfiguration
        configuration.destinationDir = destination
        configuration.clearCache = isClearCache
        return configuration
    }


    /**
     * Creates a [DefaultJBakeConfiguration] with value site.host replaced
     * by http://localhost:[server.port].
     * The server.port is read from the project properties file.
     *
     * @param destinationDir The destination directory to render and copy files to
     * @param isClearCache Whether to clear database cache or not
     * @throws JBakeExitException if loading the configuration fails
     */
    @Deprecated("use {@link #createJettyJbakeConfiguration(File, File, File, boolean)} instead")
    @Throws(JBakeExitException::class)
    fun createJettyJbakeConfiguration(sourceDir: File, destinationDir: File, isClearCache: Boolean)
        = createJettyJbakeConfiguration(sourceDir, destinationDir, null as File?, isClearCache)

    /**
     * Creates a [DefaultJBakeConfiguration] with value site.host replaced
     * by http://localhost:[server.port].
     * The server.port is read from the project properties file.
     *
     * @param destinationDir The destination directory to render and copy files to
     * @param propertiesFile The properties file for the project
     * @param isClearCache Whether to clear database cache or not
     */
    @Throws(JBakeExitException::class)
    fun createJettyJbakeConfiguration(
        sourceDir: File,
        destinationDir: File?,
        propertiesFile: File?,
        isClearCache: Boolean
    ): DefaultJBakeConfiguration {
        val conf = this.configUtil.loadConfig(sourceDir, propertiesFile) as DefaultJBakeConfiguration
        conf.destinationDir = destinationDir ?: File("./output")
        conf.clearCache = isClearCache
        conf.siteHost = "http://" + conf.serverHostname + ":" + conf.serverPort + conf.serverContextPath
        return conf
    }

    fun setEncoding(charset: String) = this.apply { configUtil.setEncoding(charset) }
}
